cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -54,76 +169,274 @@ theorem not_contains_of_empty {α β} [BEq α] (k : α) :
-  simp at h₃ ; subst h₃
-  exact List.mem_of_find?_eq_some h₂
-
-theorem mapOnValues_contains {α β γ} [LT α] [DecidableLT α] [DecidableEq α] (f : β → γ) {m : Map α β} {k : α} :
+  unfold WellFormed at wf
+  simp only [make, toList, kvs, mapOnValues, mk.injEq] at *
+  rw [wf] ; simp only ; rw [eq_comm]
+  have h₁ : Prod.map id f = (λ (x : α × β) => (x.fst, f x.snd)) := by unfold Prod.map ; simp only [id_eq]
+  simp only [← h₁, ← List.canonicalize_of_map_fst, List.canonicalize_idempotent]
+
+theorem mapOnValues_contains {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α} :
   Map.contains m k = Map.contains (Map.mapOnValues f m) k
 := by
   simp only [contains, Option.isSome]
   split
-  case h_1 h => simp [find?_mapOnValues_some f h]
-  case h_2 h => simp [find?_mapOnValues_none f h]
+  case h_1 h => simp [find?_mapOnValues_some h]
+  case h_2 h => simp [find?_mapOnValues_none h]
+
+theorem values_mapOnValues [LT α] [StrictLT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} :
+  (m.mapOnValues f).values = m.values.map f
+:= by
+  unfold mapOnValues values kvs
+  induction m.1
+  case nil => simp
+  case cons hd tl ih =>
+    simp only [List.map_cons, List.cons.injEq, true_and]
+    trivial
+
+theorem findOrErr_mapOnValues [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α} {e : Error} :
+  (m.mapOnValues f).findOrErr k e = (m.findOrErr k e).map f
+:= by
+  unfold findOrErr
+  rw [← find?_mapOnValues]
+  cases m.find? k <;> simp [Except.map]
+
+theorem findOrErr_ok_iff_find?_some [LT α] [DecidableLT α] [DecidableEq α] {m : Map α β} {k : α} {v : β} {e : Error} :
+  m.findOrErr k e = .ok v ↔ m.find? k = some v
+:= by
+  unfold findOrErr
+  cases m.find? k <;> simp
+
+theorem in_values_iff_findOrErr_ok [LT α] [DecidableLT α] [DecidableEq α] {m : Map α β} {v : β} {e : Error}
+  (wf : m.WellFormed) :
+  v ∈ m.values ↔ ∃ k, m.findOrErr k e = .ok v
+:= by
+  simp only [values, List.mem_map, findOrErr_ok_iff_find?_some]
+  constructor
+  case mp =>
+    intro h₁
+    replace ⟨⟨k, v'⟩, ⟨h₁, h₂⟩⟩ := h₁
+    simp only at h₂
+    subst v'
+    exists k
+    simp [h₁, ← in_list_some_find? wf]
+  case mpr =>
+    intro h₁
+    replace ⟨k, h₁⟩ := h₁
+    exists (k, v)
+    simp [h₁, in_list_some_find? wf, and_true]
+
+theorem in_kvs_in_mapOnValues [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α} {v : β} :
+  (k, v) ∈ m.kvs → (k, f v) ∈ (m.mapOnValues f).kvs
+:= by
+  unfold mapOnValues
+  intro h₁
+  simp only [kvs, List.mem_map, Prod.mk.injEq]
+  exists (k, v)
+
+/-! ### mapMOnValues -/
+
+theorem mapMOnValues_wf [LT α] [DecidableLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.WellFormed →
+  (m₁.mapMOnValues f = some m₂) →
+  m₂.WellFormed
+:= by
+  unfold WellFormed toList
+  intro wf h₁
+  simp only [mapMOnValues, Option.pure_def, Option.bind_eq_bind, Option.bind_eq_some,
+    Option.some.injEq] at h₁
+  replace ⟨xs, h₁, h₂⟩ := h₁
+  subst h₂
+  sorry
```

Comment:
I think this would also benefit from some List lemmas I proved internally (and would have sworn I pushed here, but clearly not!):

```
theorem mapM'_some_eq_filterMap {α β} {f : α → Option β} {xs : List α} {ys : List β} :
  List.mapM' f xs = .some ys →
  List.filterMap f xs = ys
:= by
```
It should be easier to prove with this.  You turn WellFormed into SortedBy on lists, and if the initial list was sortedBy, then its filterMap is too, which then gives the well-formedness of the result.

Reply:
#305 has a proof that does use `wf_iff_sorted`, but doesn't directly follow the above advice haha.  I should probably have reread this comment more carefully.  Let me try following this advice more directly

Reply:
Yep, this advice does lead to a good proof.  I added it as an "alternate proof" in #305.  I'm not sure which proof is better, this or #305's original proof.  #305's original proof is technically even simpler, because it relies on `mapMOnValues_preserves_keys` which does most of the work (and has a more complicated proof).  In contrast, the new alt proof following the above advice relies on `List.mapM_some_eq_filterMap` (and `List.filterMap_sortedBy`) to do most of the work.

Let's continue this conversation at https://github.com/cedar-policy/cedar-spec/pull/305#discussion_r1587795641